packages/block-library/src/column/edit.js

```diff
@@ -6,52 +6,114 @@ import classnames from 'classnames';
+						onChange={ updateWidth }
+						min={ 0 }
+						max={ 100 }
+						required
+					/>
+				</PanelBody>
+			</InspectorControls>
+			<InnerBlocks
+				templateLock={ false }
+				renderAppender={ (
+					hasChildBlocks ?
+						undefined :
+						() => <InnerBlocks.ButtonBlockAppender />
+				) }
+			/>
 		</div>
 	);
-};
+}
 
 export default compose(
-	withSelect( ( select, { clientId } ) => {
-		const { getBlockRootClientId } = select( 'core/editor' );
+	withSelect( ( select, ownProps ) => {
+		const { clientId } = ownProps;
+		const { getBlockOrder } = select( 'core/block-editor' );
 
 		return {
-			parentColumnsBlockClientId: getBlockRootClientId( clientId ),
+			hasChildBlocks: getBlockOrder( clientId ).length > 0,
 		};
 	} ),
-	withDispatch( ( dispatch, { clientId, parentColumnsBlockClientId } ) => {
+	withDispatch( ( dispatch, ownProps, registry ) => {
 		return {
-			updateAlignment( alignment ) {
-				// Update self...
-				dispatch( 'core/editor' ).updateBlockAttributes( clientId, {
-					verticalAlignment: alignment,
-				} );
+			updateAlignment( verticalAlignment ) {
+				const { clientId, setAttributes } = ownProps;
+				const { updateBlockAttributes } = dispatch( 'core/block-editor' );
+				const { getBlockRootClientId } = registry.select( 'core/block-editor' );
+
+				// Update own alignment.
+				setAttributes( { verticalAlignment } );
 
 				// Reset Parent Columns Block
-				dispatch( 'core/editor' ).updateBlockAttributes( parentColumnsBlockClientId, {
-					verticalAlignment: null,
-				} );
+				const rootClientId = getBlockRootClientId( clientId );
+				updateBlockAttributes( rootClientId, { verticalAlignment: null } );
+			},
+			updateWidth( width ) {
+				const { clientId, attributes, setAttributes } = ownProps;
+				const { updateBlockAttributes } = dispatch( 'core/block-editor' );
+				const {
+					getBlockRootClientId,
+					getBlockOrder,
+					getBlockAttributes,
+				} = registry.select( 'core/block-editor' );
+
+				// Update own width.
+				setAttributes( { width } );
+
+				// Constrain or expand siblings to account for gain or loss of
+				// total columns area.
+				const rootClientId = getBlockRootClientId( clientId );
+				const columnClientIds = getBlockOrder( rootClientId );
+				const { width: previousWidth = 100 / columnClientIds.length } = attributes;
+				const index = columnClientIds.indexOf( clientId );
+				const isLastColumn = index === columnClientIds.length - 1;
+				const increment = isLastColumn ? -1 : 1;
+				const endIndex = isLastColumn ? 0 : columnClientIds.length - 1;
+				const adjustment = ( previousWidth - width ) / Math.abs( index - endIndex );
+
+				for ( let i = index + increment; i - increment !== endIndex; i += increment ) {
+					const columnClientId = columnClientIds[ i ];
+					const { width: columnWidth = 100 / columnClientIds.length } = getBlockAttributes( columnClientId );
+					updateBlockAttributes( columnClientId, {
+						width: columnWidth + adjustment,
```

Comment:
I wonder if the computations here could be extracted to a unit-testable pure function.